pkg/skaffold/kubernetes/context/context_test.go

```diff
@@ -66,7 +93,7 @@ func TestCurrentContext(t *testing.T) {
 		config, err := CurrentConfig()
 
 		t.CheckNoError(err)
-		t.CheckDeepEqual("cluster-bar", config.CurrentContext)
+		t.CheckDeepEqual(clusterBarContext, config.CurrentContext)
 	})
 
 	testutil.Run(t, "invalid context", func(t *testutil.T) {
```

Comment:
This test fails now (and didn't beforehand). Should have included that case...
```go
testutil.Run(t, "change context after first execution", func(t *testutil.T) {
		resetKubeConfig(t, validKubeConfig)

		cfg, err := GetRestClientConfig()
		kubeContext = clusterBarContext

		t.CheckNoError(err)
		t.CheckDeepEqual("https://bar.com", cfg.Host)
	})
```

Reply:
This testcase fails on `master` too. 
I think you wanted 

```
	testutil.Run(t, "change context after first execution", func(t *testutil.T) {
		resetKubeConfig(t, validKubeConfig)

		cfg, err := GetRestClientConfig()
		kubeContext = clusterBarContext
		cfg, err = GetRestClientConfig()

		t.CheckNoError(err)
		t.CheckDeepEqual("https://bar.com", cfg.Host)
	})
```

This requirement is future looking - currently we deal with a single kubeContext per skaffold execution. Your argument around kubectl vs client-go becoming inconsistent does not apply in the current setting: we call `UseKubeContext()` at the very beginning of the skaffold process, and kubectl CLI is instantiated from the `RunContext.KubeContext`. I would even close down the `UseKubeContext()` method with a `sync.Once` to allow only a single modification of that private field. 

Maybe in the future we could have a situation where we need to interact with multiple contexts in multiple deployments. Assuming that the kubeconfig is the same in this case and is cached similarly to what we have now - we just use different kubeContexts from it - in that case we should be able to create both CLIs (which is easy currently) and REST clients that points to the same context in the kubeconfig...but then we wouldn't drive it through _changing_ the global kubeContext variable, but rather by creating  a `GetRestClientConfig(kubeContext string)` method. And that would be broken, given my solution points to the `DeferredLoading...ClientConfig` which caches the actual kubeconfig after first call and does not allow for overrides - i.e. we can cache the kubeconfig at the beginning of the process but we won't be able to pick kubecontexts from it. 

Now your original solution works with picking contexts from the cached kubeconfig. However it lacks the logic for in-cluster situations. Now, thinking about it - we should think about he in-cluster config as _another context_. Thus here's what I'll implement: 

- use the override logic you're using with the current kubecontext (which might be empty)
- if kubecontext is empty and the resulting kubeconfig is empty - try the in-cluster rest-config

Reply:
> This testcase fails on master too.

:sweat_smile: